stor/dx.py

```diff
@@ -385,129 +1067,161 @@ def exists(self):
-                                          project=self.canonical_project).describe()
-            else:
-                self.__stat = dxpy.DXProject(dxid=self.canonical_project).describe()
-        return self.__stat
+        Raises:
+            DuplicateError: If project or resource is not unique
+            NotFoundError: When the project or resource cannot be found
+            ValueError: If path is folder path
+        """
+        if not self.resource:
+            return dxpy.DXProject(dxid=self.canonical_project).describe()
+        return dxpy.DXFile(dxid=self.canonical_resource,
+                           project=self.canonical_project).describe()
 
 
 class DXVirtualPath(DXPath):
-    """Class Handler for DXPath of form 'dx://project-{ID}:/a/b/c' or 'dx://a/b/c'"""
+    """Class Handler for DXPath of form 'dx://project-{dxID}:/a/b/c' or 'dx://a:/b/c'"""
 
     @property
     def virtual_project(self):
+        """Returns the virtual name of the project associated with the DXVirtualPath"""
         if utils.is_valid_dxid(self.project, 'project'):
-            return dxpy.DXProject(dxid=self.project).name
+            with _propagate_dx_exceptions():
+                return dxpy.DXProject(dxid=self.project).name
         return self.project
 
     @property
     def virtual_resource(self):
+        """Returns the human readable path of the resource associated with the DXVirtualPath"""
         return self.resource
 
     @property
     def virtual_path(self):
+        """Returns the DXVirtualPath instance"""
         return self
 
     @cached_property
     def canonical_project(self):
-        """Returns the unique project that matches the name that user has view access to.
-        If no match is found, returns None
+        """Returns the ID of the unique project that matches the name that user has at least
+        view access to. If no match is found, returns None
 
         Raises:
-            DuplicateProjectError - if project name is not unique on DX platform
-            NotFoundError - If project name doesn't exist on DNAnexus
+            DuplicateProjectError: If project name is not unique on DX platform
+            NotFoundError: If project name doesn't exist on DNAnexus
         """
         if utils.is_valid_dxid(self.project, 'project'):
             return self.project
-        else:
+
+        with _propagate_dx_exceptions():
             try:
                 proj_dict = dxpy.find_one_project(
                     name=self.project, level='VIEW', zero_ok=True, more_ok=False)
-                if proj_dict is None:
-                    raise ProjectNotFoundError('No projects were found with given name ({})'
-                                               .format(self.project))
-                return proj_dict['id']
-            except DXError as e:
+            except DXSearchError as e:
                 raise DuplicateProjectError('Duplicate projects were found with given name ({})'
```

Comment:
```suggestion
                raise DuplicateProjectError('Multiple projects were found with given name ({})'
```

Reply:
duplicate isn't the right word, because it means the same, really we could call this `MultipleObjectSameNameError`? and could reuse for both? (possibly with subclasses for MultipleProjects and MultipleObjects).

What does dxpy return for this?

Reply:
Since, we have `more_ok=False`, `dxpy` also raises a `DXSearchError: Expected one result, but found more` for this which is caught here. If `more_ok=True`, it asks user to choose between same named projects. Changed the error, and reused for both. DIdn't subclass.